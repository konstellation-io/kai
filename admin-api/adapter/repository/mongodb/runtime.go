package mongodb

import (
	"context"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/adapter/config"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/domain/entity"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/domain/usecase/logging"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"time"
)

type RuntimeRepoMongoDB struct {
	cfg        *config.Config
	logger     logging.Logger
	collection *mongo.Collection
}

func NewRuntimeRepoMongoDB(cfg *config.Config, logger logging.Logger, client *mongo.Client) *RuntimeRepoMongoDB {
	collection := client.Database(cfg.MongoDB.DBName).Collection("runtimes")
	return &RuntimeRepoMongoDB{
		cfg,
		logger,
		collection,
	}
}

func (r *RuntimeRepoMongoDB) Create(name string, userID string) (*entity.Runtime, error) {
	runtime := &entity.Runtime{
		ID:           primitive.NewObjectID().Hex(),
		Name:         name,
		CreationDate: time.Now().UTC(),
		Owner:        userID,
	}
	res, err := r.collection.InsertOne(context.Background(), runtime)
	if err != nil {
		return nil, err
	}

	runtime.ID = res.InsertedID.(string)
	return runtime, nil
}
