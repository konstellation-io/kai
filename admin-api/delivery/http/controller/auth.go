package controller

import (
	"github.com/dgrijalva/jwt-go"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/domain/usecase/logging"
	"net/http"
	"time"

	"github.com/labstack/echo"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/adapter/config"
	"gitlab.com/konstellation/konstellation-ce/kre/admin-api/domain/usecase"
)

type signInInput struct {
	Email string `json:"email" validate:"required,email"`
}

type signinVerifyInput struct {
	VerificationCode string `json:"verificationCode" validate:"required"`
}

type AuthController struct {
	cfg               *config.Config
	logger            logging.Logger
	authInteractor    *usecase.AuthInteractor
	settingInteractor *usecase.SettingInteractor
}

func NewAuthController(
	cfg *config.Config,
	logger logging.Logger,
	authInteractor *usecase.AuthInteractor,
	settingInteractor *usecase.SettingInteractor,
) *AuthController {
	return &AuthController{
		cfg,
		logger,
		authInteractor,
		settingInteractor,
	}
}

func (a *AuthController) SignIn(c echo.Context) error {
	input := new(signInInput)
	if err := c.Bind(input); err != nil {
		return HTTPErrInvalidJSON
	}

	if err := c.Validate(input); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, newResponseValidationError(err))
	}

	verificationCodeDurationInMinutes := a.cfg.Auth.VerificationCodeDurationInMinutes
	if err := a.authInteractor.SignIn(input.Email, verificationCodeDurationInMinutes); err != nil {
		switch err {
		case usecase.ErrDomainNotAllowed:
			return HTTPErrDomainNotAllowed
		default:
			a.logger.Error(err.Error())
			return HTTPErrUnexpected
		}
	}

	return c.JSON(http.StatusOK, map[string]interface{}{"message": "Email sent to the user."})
}

func (a *AuthController) SignInVerify(c echo.Context) error {
	input := new(signinVerifyInput)
	if err := c.Bind(input); err != nil {
		return HTTPErrInvalidJSON
	}

	if err := c.Validate(input); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, newResponseValidationError(err))
	}

	a.logger.Info("Verifying authorization code.")
	userId, err := a.authInteractor.VerifyCode(input.VerificationCode)
	if err != nil {
		switch err {
		case usecase.ErrExpiredVerificationCode:
		case usecase.ErrVerificationCodeNotFound:
			return HTTPErrVerificationCodeNotFound
		default:
			a.logger.Error(err.Error())
			return HTTPErrUnexpected
		}
	}

	a.logger.Info("Generating JWT token.")
	setting, err := a.settingInteractor.Get()
	if err != nil {
		return err
	}

	ttl := time.Duration(setting.SessionLifetimeInDays) * 24 * time.Hour
	expirationTime := time.Now().Add(ttl)

	claims := jwt.StandardClaims{
		Subject:   userId,
		ExpiresAt: expirationTime.Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)

	jwtKey := []byte(a.cfg.Auth.JWTSignSecret)
	jwtToken, err := token.SignedString(jwtKey)
	if err != nil {
		return err
	}

	a.logger.Info("Set cookie containing the generated JWT token.")
	cookie := new(http.Cookie)
	cookie.Name = "token"
	cookie.Value = jwtToken
	cookie.Expires = expirationTime
	cookie.Path = "/"
	cookie.Secure = a.cfg.Auth.SecureCookie
	cookie.HttpOnly = true
	c.SetCookie(cookie)

	return c.JSON(http.StatusOK, map[string]interface{}{"message": "Login success"})
}

func (a *AuthController) Logout(c echo.Context) error {
	user := c.Get("user").(*jwt.Token)
	claims := user.Claims.(jwt.MapClaims)
	userID := claims["sub"].(string)

	a.logger.Info("Logout for user " + userID)

	expirationTime := time.Now()
	cookie := new(http.Cookie)
	cookie.Name = "token"
	cookie.Value = "deleted"
	cookie.Expires = expirationTime
	cookie.Path = "/"
	cookie.Secure = a.cfg.Auth.SecureCookie
	cookie.HttpOnly = true
	c.SetCookie(cookie)

	return c.JSON(http.StatusOK, map[string]interface{}{"message": "Logout success"})
}
