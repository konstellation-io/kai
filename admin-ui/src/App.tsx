import React from 'react';
import { Router } from 'react-router-dom';
import { Route, Redirect, Switch } from 'react-router';
import history from './history';
import * as PAGES from './constants/routes';

import Login from './pages/Login/Login';
import VerifyEmail from './pages/VerifyEmail/VerifyEmail';
import MagicLink from './pages/MagicLink/MagicLink';
import Dashboard from './pages/Dashboard/Dashboard';
import Runtime from './pages/Runtime/Runtime';
import Settings from './pages/Settings/Settings';
import AddRuntime from './pages/AddRuntime/AddRuntime';
import AddVersion from './pages/AddVersion/AddVersion';

export function Routes() {
  return (
    <>
      <Switch>
        <Redirect exact from={PAGES.HOME} to={PAGES.DASHBOARD} />
        <Redirect exact from={PAGES.SETTINGS} to={PAGES.SETTINGS_GENERAL} />
        <Redirect exact from={PAGES.RUNTIME} to={PAGES.RUNTIME_STATUS} />

        <Route exact path={PAGES.LOGIN} component={Login} />
        <Route exact path={PAGES.VERIFY_EMAIL} component={VerifyEmail} />
        <Route exact path={PAGES.MAGIC_LINK} component={MagicLink} />

        <Route path={PAGES.NEW_RUNTIME} component={AddRuntime} />
        <Route path={PAGES.NEW_VERSION} component={AddVersion} />

        <Route exact path={PAGES.DASHBOARD} component={Dashboard} />
        <Route path={PAGES.RUNTIME} component={Runtime} />
        <Route path={PAGES.SETTINGS} component={Settings} />
      </Switch>
    </>
  );
}

function App() {
  return (
    <div className="app">
      <Router history={history}>
        <Routes />
      </Router>
    </div>
  );
}

export default App;
