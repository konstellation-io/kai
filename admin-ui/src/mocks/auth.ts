import { GET_USER_EMAIL } from '../components/Header/Header.graphql';

export const usernameMock = {
  request: {
    query: GET_USER_EMAIL
  },
  result: {
    data: {
      me: {
        email: 'user@konstellation.com'
      }
    }
  }
};

export const unauthorizedUsernameMock = {
  request: {
    query: GET_USER_EMAIL
  },
  result: {
    data: {
      me: {
        email: 'unauthorizedUser@konstellation.com'
      }
    }
  },
  error: new Error('Unauthorized user')
};
