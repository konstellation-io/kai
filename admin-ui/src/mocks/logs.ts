import { GET_LOGS } from '../graphql/client/queries/getLogs.graphql';

export const currentLogPanelMock = {
  request: {
    query: GET_LOGS
  },
  result: {
    data: {
      runtimeId: 'runtimeIdMock',
      nodeId: 'nodeIdMock',
      nodeName: 'nodeNameMock'
    }
  }
};
