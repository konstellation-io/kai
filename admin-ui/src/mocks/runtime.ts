import { GET_DASHBOARD } from '../pages/Dashboard/Dashboard.graphql';
import { ADD_RUNTIME } from '../pages/AddRuntime/AddRuntime.graphql';

export const dashboardMock = {
  request: {
    query: GET_DASHBOARD
  },
  result: {
    data: {
      dashboard: {
        runtimes: [
          {
            id: '00001',
            status: 'ACTIVE',
            name: 'Some Name',
            creationDate: '2018-01-02'
          },
          {
            id: '00002',
            status: 'ACTIVE',
            name: 'Some Other Name',
            creationDate: '2018-02-03'
          }
        ],
        alerts: [
          {
            id: 'id0',
            type: 'error',
            message: 'some message',
            runtime: { id: '00001' }
          },
          {
            id: 'id1',
            type: 'error',
            message: 'some message 2',
            runtime: { id: '00001' }
          }
        ]
      }
    }
  }
};

export const addRuntimeMock = {
  request: {
    query: ADD_RUNTIME,
    variables: { name: 'New Runtime' }
  },
  result: {
    data: {
      createRuntime: {
        success: true,
        message: 'some message'
      }
    }
  }
};
