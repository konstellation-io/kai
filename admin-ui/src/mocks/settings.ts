import {
  GET_DOMAINS,
  GET_EXPIRATION_TIME,
  ADD_ALLOWED_DOMAIN,
  UPDATE_EXPIRATION_TIME
} from '../pages/Settings/Settings.graphql';
import { GET_USERS_ACTIVITY } from '../components/UserActivityList/UserActivityList.graphql';

export const domainMock = {
  request: {
    query: GET_DOMAINS
  },
  result: {
    data: {
      settings: {
        authAllowedDomains: ['domain.1', 'domain.2', 'domain.3.sample']
      }
    }
  }
};

export const usersActivityMock = {
  request: {
    query: GET_USERS_ACTIVITY
  },
  result: {
    data: {
      usersActivity: [
        {
          user: 'user1@domain.com',
          message: 'some message 1',
          date: '2019-01-01'
        },
        {
          user: 'user2@domain.com',
          message: 'some message 2',
          date: '2019-01-02'
        },
        {
          user: 'user3@domain.com',
          message: 'some message 3',
          date: '2019-01-03'
        }
      ]
    }
  }
};

export const expirationTimeMock = {
  request: {
    query: GET_EXPIRATION_TIME
  },
  result: {
    data: {
      settings: {
        cookieExpirationTime: 45
      }
    }
  }
};

export const addAllowedDomainMock = {
  request: {
    query: ADD_ALLOWED_DOMAIN,
    variables: { domainName: 'intelygenz.com' }
  },
  result: {
    data: {
      addAllowedDomain: {
        authAllowedDomains: [
          'domain.1',
          'domain.2',
          'domain.3.sample',
          'intelygenz.com'
        ]
      }
    }
  }
};

export const removeAllowedDomainMock = {
  request: {
    query: ADD_ALLOWED_DOMAIN,
    variables: { domainName: 'domain.3.sample' }
  },
  result: {
    data: {
      removeAllowedDomain: {
        authAllowedDomains: ['domain.1', 'domain.2']
      }
    }
  }
};

export const updateExpirationTime = {
  request: {
    query: UPDATE_EXPIRATION_TIME,
    variables: { input: { cookieExpirationTime: 10 } }
  },
  result: {
    data: {
      setSettings: {
        cookieExpirationTime: 12
      }
    }
  }
};
