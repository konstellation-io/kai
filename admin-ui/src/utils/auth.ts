import Cookies from 'js-cookie';

export const JWT_COOKIE = '__jwt';

export const isUserAuthenticated = () => {
  const jwt = Cookies.get(JWT_COOKIE);

  return jwt;
};
