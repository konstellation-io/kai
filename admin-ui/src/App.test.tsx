import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import ReactDOM from 'react-dom';
import App, { Routes } from './App';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom/extend-expect';

import { MockedProvider } from '@apollo/react-testing';
import wait from 'waait';
import { dashboardMock } from './mocks/runtime';
import { usernameMock, unauthorizedUsernameMock } from './mocks/auth';

const mocks = [dashboardMock, usernameMock];
const mocksUnauthorizedUser = [dashboardMock, unauthorizedUsernameMock];

afterEach(cleanup);

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <App />
      </MemoryRouter>
    </MockedProvider>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

// I cannot test login redirection, redirection logic is located inside Apollo client config,
// But I can check for apollo errors when making a request
it('it shows spash screen when user is not logged in', async () => {
  const { getByTestId } = render(
    <MockedProvider mocks={mocksUnauthorizedUser} addTypename={false}>
      <MemoryRouter>
        <Routes />
      </MemoryRouter>
    </MockedProvider>
  );

  await act(async () => {
    await wait(0);
  });

  expect(getByTestId('splashscreen')).toBeInTheDocument();
});

it('it shows dashboard page on home URL when logged', () => {
  const { getByTestId } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <Routes />
      </MemoryRouter>
    </MockedProvider>
  );

  expect(getByTestId('dashboardContainer')).toBeInTheDocument();
});
