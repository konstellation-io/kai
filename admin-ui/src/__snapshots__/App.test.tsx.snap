// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="splash"
      data-testid="splashscreen"
    />
    <div
      class="container"
      data-testid="dashboardContainer"
    >
      <div
        class="container"
      >
        <div
          class="container"
        >
          <div
            class="button active"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-regular"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 7.77L18.39 18H5.61L12 7.77M12 4L2 20h20L12 4z"
              />
            </svg>
          </div>
        </div>
        <div
          class="container"
        >
          <div
            class="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-regular"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 1L3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4zm-2 16l-4-4 1.41-1.41L10 14.17l6.59-6.59L18 9l-8 8z"
              />
            </svg>
          </div>
        </div>
        <div
          class="container"
        >
          <div
            class="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-regular"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 12c1.38 0 2.49-1.12 2.49-2.5S17.88 7 16.5 7C15.12 7 14 8.12 14 9.5s1.12 2.5 2.5 2.5zM9 11c1.66 0 2.99-1.34 2.99-3S10.66 5 9 5C7.34 5 6 6.34 6 8s1.34 3 3 3zm7.5 3c-1.83 0-5.5.92-5.5 2.75V19h11v-2.25c0-1.83-3.67-2.75-5.5-2.75zM9 13c-2.33 0-7 1.17-7 3.5V19h7v-2.25c0-.85.33-2.34 2.37-3.47C10.5 13.1 9.66 13 9 13z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class="hexagons"
        >
          <div
            class="contentBox"
            data-testid="spinner"
          >
            <div
              class="container"
              style="width: 40px; height: 40px;"
            >
              <div
                class="sun"
                style="background-color: white;"
              />
              <div
                class="orbit p1"
                style="border-color: white;"
              />
              <div
                class="planet_container p1"
              >
                <div
                  class="planet p1"
                  style="background-color: white;"
                >
                  <div
                    class="planet_container p4"
                    data-testid="spinner-planet"
                  >
                    <div
                      class="planet p4"
                      style="background-color: white;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="orbit p2"
                style="border-color: white;"
              />
              <div
                class="planet_container p2"
              >
                <div
                  class="planet p2"
                  style="background-color: white;"
                />
              </div>
              <div
                class="planet_container p3"
              >
                <div
                  class="planet p3"
                  style="background-color: white;"
                />
              </div>
              <div
                class="orbit p5"
                style="border-color: white;"
              />
              <div
                class="planet_container p5"
              >
                <div
                  class="planet p5"
                  style="background-color: white;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
